File: scripts/generate-melodies.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const OUTPUT_DIR = path.join(__dirname, '..', 'public', 'sounds');
const SAMPLE_RATE = 44100;
const DURATION = 5; // seconds

// Note frequencies
const NOTES = {
    'C4': 261.63, 'D4': 293.66, 'E4': 329.63, 'F4': 349.23, 'G4': 392.00, 'A4': 440.00, 'B4': 493.88,
    'C5': 523.25, 'D5': 587.33, 'E5': 659.25, 'F5': 698.46, 'G5': 783.99, 'A5': 880.00,
    'REST': 0
};

// Simple melodies (approximate for 5 seconds)
// Format: [Note, Duration (beats)]
// Tempo: approx 120 BPM (0.5s per beat) -> Actually let's make it faster notes for 5s clip
// We'll use 0.25s as base unit
const MELODIES = {
    'twinkle': [ // Twinkle Twinkle Little Star
        ['C4', 1], ['C4', 1], ['G4', 1], ['G4', 1], ['A4', 1], ['A4', 1], ['G4', 2],
        ['F4', 1], ['F4', 1], ['E4', 1], ['E4', 1], ['D4', 1], ['D4', 1], ['C4', 2]
    ],
    'happy': [ // Happy Birthday
        ['C4', 0.75], ['C4', 0.25], ['D4', 1], ['C4', 1], ['F4', 1], ['E4', 2],
        ['C4', 0.75], ['C4', 0.25], ['D4', 1], ['C4', 1], ['G4', 1], ['F4', 2]
    ],
    'wheels': [ // Wheels on the Bus
        ['C4', 1], ['F4', 1], ['F4', 0.5], ['F4', 0.5], ['F4', 1], ['A4', 1], ['C5', 1], ['A4', 1], ['F4', 2],
        ['G4', 1], ['E4', 1], ['C4', 2]
    ],
    'bingo': [ // Bingo
        ['G4', 1], ['C5', 1], ['C5', 1], ['C5', 0.5], ['C5', 0.5], ['A4', 1], ['A4', 1],
        ['G4', 0.5], ['G4', 0.5], ['F4', 1], ['F4', 1], ['E4', 1], ['D4', 1], ['C4', 2]
    ],
    'old_macdonald': [ // Old MacDonald
        ['G4', 1], ['G4', 1], ['G4', 1], ['D4', 1], ['E4', 1], ['E4', 1], ['D4', 2],
        ['B4', 1], ['B4', 1], ['A4', 1], ['A4', 1], ['G4', 2]
    ],
    'itsy_bitsy': [ // Itsy Bitsy Spider
        ['G4', 1], ['C4', 1], ['C4', 1], ['D4', 1], ['E4', 1], ['E4', 1], ['E4', 1], ['D4', 0.5], ['C4', 0.5],
        ['D4', 1], ['E4', 1], ['C4', 2]
    ],
    'mary_lamb': [ // Mary Had a Little Lamb
        ['E4', 1], ['D4', 1], ['C4', 1], ['D4', 1], ['E4', 1], ['E4', 1], ['E4', 2],
        ['D4', 1], ['D4', 1], ['D4', 2], ['E4', 1], ['G4', 1], ['G4', 2]
    ],
    'row_boat': [ // Row Your Boat
        ['C4', 1.5], ['C4', 1.5], ['C4', 1], ['D4', 0.5], ['E4', 1.5],
        ['E4', 1], ['D4', 0.5], ['E4', 1], ['F4', 0.5], ['G4', 3]
    ],
    'abc': [ // ABC Song
        ['C4', 1], ['C4', 1], ['G4', 1], ['G4', 1], ['A4', 1], ['A4', 1], ['G4', 2],
        ['F4', 1], ['F4', 1], ['E4', 1], ['E4', 1], ['D4', 1], ['D4', 1], ['C4', 2] // Same as Twinkle roughly
    ],
    'london_bridge': [ // London Bridge
        ['G4', 1.5], ['A4', 0.5], ['G4', 1], ['F4', 1], ['E4', 1], ['F4', 1], ['G4', 2],
        ['D4', 1], ['E4', 1], ['F4', 2], ['E4', 1], ['F4', 1], ['G4', 2]
    ]
};

// WAV Header functions
function writeString(view, offset, string) {
    for (let i = 0; i < string.length; i++) {
        view.setUint8(offset + i, string.charCodeAt(i));
    }
}

function generateSilence(duration) {
    return new Float32Array(SAMPLE_RATE * duration);
}

function generateTone(freq, duration, type = 'sine') {
    const numSamples = Math.floor(SAMPLE_RATE * duration);
    const buffer = new Float32Array(numSamples);
    const envelope = 0.1; // Attack/Release time in seconds
    const attackSamples = SAMPLE_RATE * 0.05;
    const decaySamples = SAMPLE_RATE * 0.4;

    for (let i = 0; i < numSamples; i++) {
        const t = i / SAMPLE_RATE;
        let value = 0;

        // Sine wave
        value = Math.sin(2 * Math.PI * freq * t);

        // Add some harmonics for "Music Box" sound (like a triangle/bell)
        if (freq > 0) {
            value += 0.5 * Math.sin(2 * Math.PI * freq * 2 * t);
            value += 0.25 * Math.sin(2 * Math.PI * freq * 3 * t);
        }

        // Apply Envelope (Attack, Decay, Sustain, Release)
        let amp = 0.5; // Master volume
        if (i < attackSamples) {
            amp *= i / attackSamples;
        } else {
            // Natural exponential decay for bell-like sound
            const decayProgress = (i - attackSamples) / SAMPLE_RATE;
            amp *= Math.exp(-3 * decayProgress);
        }

        buffer[i] = value * amp;
    }
    return buffer;
}

function bufferToWav(buffer) {
    const numChannels = 1;
    const bitDepth = 16;
    const bytesPerSample = bitDepth / 8;
    const blockAlign = numChannels * bytesPerSample;
    const byteRate = SAMPLE_RATE * blockAlign;
    const dataSize = buffer.length * blockAlign;
    const headerSize = 44;
    const totalSize = headerSize + dataSize;

    const arrayBuffer = new ArrayBuffer(totalSize);
    const view = new DataView(arrayBuffer);

    // RIFF chunk
    writeString(view, 0, 'RIFF');
    view.setUint32(4, 36 + dataSize, true);
    writeString(view, 8, 'WAVE');

    // fmt chunk
    writeString(view, 12, 'fmt ');
    view.setUint32(16, 16, true); // Subchunk1Size (16 for PCM)
    view.setUint16(20, 1, true); // AudioFormat (1 for PCM)
    view.setUint16(22, numChannels, true);
    view.setUint32(24, SAMPLE_RATE, true);
    view.setUint32(28, byteRate, true);
    view.setUint16(32, blockAlign, true);
    view.setUint16(34, bitDepth, true);

    // data chunk
    writeString(view, 36, 'data');
    view.setUint32(40, dataSize, true);

    // Write samples
    const pcmData = new Int16Array(arrayBuffer, 44, buffer.length);
    for (let i = 0; i < buffer.length; i++) {
        // Clip and scale to 16-bit
        let s = Math.max(-1, Math.min(1, buffer[i]));
        pcmData[i] = s < 0 ? s * 0x8000 : s * 0x7FFF;
    }

    return Buffer.from(arrayBuffer);
}

// Generate all melodies
if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

Object.entries(MELODIES).forEach(([key, notes]) => {
    let fullBuffer = new Float32Array(0);

    // Calculate timing to fit roughly in 5 seconds or loop
    // Base unit duration
    const beatDuration = 0.35; // seconds

    notes.forEach(([note, beats]) => {
        const freq = NOTES[note];
        const duration = beats * beatDuration;
        const toneBuffer = generateTone(freq, duration);

        // Concatenate
        const newBuffer = new Float32Array(fullBuffer.length + toneBuffer.length);
        newBuffer.set(fullBuffer);
        newBuffer.set(toneBuffer, fullBuffer.length);
        fullBuffer = newBuffer;
    });

    // Pad or trim to exactly 5 seconds
    const targetSamples = SAMPLE_RATE * DURATION;
    let finalBuffer = new Float32Array(targetSamples);

    if (fullBuffer.length < targetSamples) {
        // Loop if too short
        let offset = 0;
        while (offset < targetSamples) {
            const chunk = Math.min(fullBuffer.length, targetSamples - offset);
            finalBuffer.set(fullBuffer.subarray(0, chunk), offset);
            offset += chunk;
        }
    } else {
        finalBuffer.set(fullBuffer.subarray(0, targetSamples));
    }

    // Apply a fade out at the very end to avoid clicking
    const fadeOutSamples = 1000;
    for (let i = 0; i < fadeOutSamples; i++) {
        finalBuffer[targetSamples - 1 - i] *= (i / fadeOutSamples);
    }

    const wavData = bufferToWav(finalBuffer);
    const filePath = path.join(OUTPUT_DIR, `${key}.wav`);
    fs.writeFileSync(filePath, wavData);
    console.log(`✓ Generated ${key}.wav`);
});

console.log('✅ Generated all melody files!');

File: scripts/generate-melodies.py
#!/usr/bin/env python3
"""
Generate recognizable children's song melodies as WAV files.
Uses pure sine waves with envelope shaping for a clean, musical sound.
No external dependencies - uses only Python builtins.
"""

import wave
import struct
import math
import os

SAMPLE_RATE = 44100
SOUNDS_DIR = "public/sounds"

# Note frequencies (Hz)
NOTES = {
    'C3': 130.81, 'D3': 146.83, 'E3': 164.81, 'F3': 174.61, 'G3': 196.00,
    'A3': 220.00, 'B3': 246.94,
    'C4': 261.63, 'C#4': 277.18, 'D4': 293.66, 'D#4': 311.13,
    'E4': 329.63, 'F4': 349.23, 'F#4': 369.99, 'G4': 392.00,
    'G#4': 415.30, 'A4': 440.00, 'A#4': 466.16, 'B4': 493.88,
    'C5': 523.25, 'D5': 587.33, 'E5': 659.25, 'F5': 698.46, 'G5': 783.99,
    'R': 0,  # Rest
}


def generate_tone(freq, duration, volume=0.6):
    """Generate a sine wave tone with smooth attack/release envelope."""
    n_samples = int(SAMPLE_RATE * duration)
    samples = []
    
    # Attack and release times (in samples)
    attack = int(SAMPLE_RATE * 0.02)   # 20ms attack
    release = int(SAMPLE_RATE * 0.04)  # 40ms release
    
    for i in range(n_samples):
        # Envelope
        if i < attack:
            env = i / attack
        elif i > n_samples - release:
            env = (n_samples - i) / release
        else:
            env = 1.0
        
        if freq == 0:  # Rest
            samples.append(0)
        else:
            # Mix fundamental with a softer octave for a richer sound
            val = math.sin(2 * math.pi * freq * i / SAMPLE_RATE) * 0.7
            val += math.sin(2 * math.pi * freq * 2 * i / SAMPLE_RATE) * 0.2
            val += math.sin(2 * math.pi * freq * 3 * i / SAMPLE_RATE) * 0.1
            samples.append(val * env * volume)
    
    return samples


def generate_melody(notes_sequence, filename):
    """
    Generate a WAV file from a sequence of (note, duration) tuples.
    """
    all_samples = []
    for note, duration in notes_sequence:
        freq = NOTES[note]
        all_samples.extend(generate_tone(freq, duration))
    
    # Add 0.3s silence at end
    all_samples.extend([0] * int(SAMPLE_RATE * 0.3))
    
    # Normalize
    max_val = max(abs(s) for s in all_samples) if all_samples else 1
    if max_val > 0:
        all_samples = [s / max_val * 0.9 for s in all_samples]
    
    # Write WAV
    filepath = os.path.join(SOUNDS_DIR, filename)
    with wave.open(filepath, 'w') as wf:
        wf.setnchannels(1)
        wf.setsampwidth(2)
        wf.setframerate(SAMPLE_RATE)
        for s in all_samples:
            wf.writeframes(struct.pack('<h', int(s * 32767)))
    
    duration_s = len(all_samples) / SAMPLE_RATE
    print(f"✓ {filename} ({duration_s:.1f}s)")


# ==================== MELODIES ====================
# Each melody is a list of (note_name, duration_in_seconds)

D = 0.35   # Default note duration
H = 0.70   # Half note (long)
Q = 0.25   # Quick note
E = 0.18   # Eighth note

# 1. Twinkle Twinkle Little Star
twinkle = [
    ('C4', D), ('C4', D), ('G4', D), ('G4', D), ('A4', D), ('A4', D), ('G4', H),
    ('F4', D), ('F4', D), ('E4', D), ('E4', D), ('D4', D), ('D4', D), ('C4', H),
    ('G4', D), ('G4', D), ('F4', D), ('F4', D), ('E4', D), ('E4', D), ('D4', H),
    ('G4', D), ('G4', D), ('F4', D), ('F4', D), ('E4', D), ('E4', D), ('D4', H),
]

# 2. Happy Birthday
happy = [
    ('C4', Q), ('C4', E), ('D4', D), ('C4', D), ('F4', D), ('E4', H),
    ('C4', Q), ('C4', E), ('D4', D), ('C4', D), ('G4', D), ('F4', H),
    ('C4', Q), ('C4', E), ('C5', D), ('A4', D), ('F4', D), ('E4', D), ('D4', H),
    ('A#4', Q), ('A#4', E), ('A4', D), ('F4', D), ('G4', D), ('F4', H),
]

# 3. Wheels on the Bus
wheels = [
    ('C4', D), ('E4', D), ('G4', D), ('G4', D), ('G4', H),
    ('A4', D), ('G4', H), ('R', Q),
    ('F4', D), ('E4', D), ('D4', D), ('D4', D), ('D4', H),
    ('E4', D), ('D4', H), ('R', Q),
    ('C4', D), ('E4', D), ('G4', D), ('G4', D), ('G4', D),
    ('A4', D), ('G4', D), ('F4', D), ('E4', D), ('D4', D), ('C4', H),
]

# 4. B-I-N-G-O
bingo = [
    ('G4', D), ('G4', D), ('E4', D), ('G4', D), ('A4', D), ('A4', H),
    ('G4', D), ('G4', D), ('E4', D), ('G4', D), ('F#4', H),
    ('E4', D), ('E4', D), ('D4', D), ('E4', D), ('F#4', D), ('F#4', D),
    ('E4', D), ('E4', D), ('D4', D), ('E4', D), ('D4', H),
]

# 5. Old MacDonald Had a Farm
old_macdonald = [
    ('G4', D), ('G4', D), ('G4', D), ('D4', D), ('E4', D), ('E4', D), ('D4', H),
    ('B4', D), ('B4', D), ('A4', D), ('A4', D), ('G4', H), ('R', Q),
    ('D4', D), ('G4', D), ('G4', D), ('G4', D), ('D4', D), ('E4', D), ('E4', D), ('D4', H),
    ('B4', D), ('B4', D), ('A4', D), ('A4', D), ('G4', H),
]

# 6. Itsy Bitsy Spider
itsy_bitsy = [
    ('G4', Q), ('C4', D), ('C4', E), ('C4', D), ('D4', D), ('E4', D), ('E4', H),
    ('E4', D), ('D4', D), ('C4', D), ('D4', D), ('E4', D), ('C4', H),
    ('E4', D), ('E4', D), ('F4', D), ('G4', H),
    ('G4', D), ('F4', D), ('E4', D), ('F4', D), ('G4', D), ('E4', H),
    ('C4', D), ('C4', D), ('D4', D), ('E4', H),
    ('E4', D), ('D4', D), ('C4', D), ('D4', D), ('E4', D), ('C4', H),
]

# 7. Mary Had a Little Lamb
mary_lamb = [
    ('E4', D), ('D4', D), ('C4', D), ('D4', D), ('E4', D), ('E4', D), ('E4', H),
    ('D4', D), ('D4', D), ('D4', H), ('E4', D), ('G4', D), ('G4', H),
    ('E4', D), ('D4', D), ('C4', D), ('D4', D), ('E4', D), ('E4', D), ('E4', D), ('E4', D),
    ('D4', D), ('D4', D), ('E4', D), ('D4', D), ('C4', H),
]

# 8. Row Row Row Your Boat
row_boat = [
    ('C4', D), ('C4', D), ('C4', Q), ('D4', E), ('E4', D),
    ('E4', Q), ('D4', E), ('E4', Q), ('F4', E), ('G4', H),
    ('C5', E), ('C5', E), ('C5', E), ('G4', E), ('G4', E), ('G4', E),
    ('E4', E), ('E4', E), ('E4', E), ('C4', E), ('C4', E), ('C4', E),
    ('G4', Q), ('F4', E), ('E4', Q), ('D4', E), ('C4', H),
]

# 9. ABC Song (same melody as Twinkle)
abc = [
    ('C4', D), ('C4', D), ('G4', D), ('G4', D), ('A4', D), ('A4', D), ('G4', H),
    ('F4', D), ('F4', D), ('E4', D), ('E4', Q), ('D4', Q), ('D4', Q), ('D4', Q), ('C4', H),
    ('G4', D), ('G4', Q), ('F4', H), ('E4', D), ('E4', D), ('D4', H),
    ('G4', Q), ('G4', Q), ('G4', Q), ('F4', H), ('E4', D), ('E4', D), ('D4', H),
    ('C4', D), ('C4', D), ('G4', D), ('G4', D), ('A4', D), ('A4', D), ('G4', H),
    ('F4', D), ('F4', D), ('E4', D), ('E4', D), ('D4', D), ('D4', D), ('C4', H),
]

# 10. London Bridge Is Falling Down
london_bridge = [
    ('G4', D), ('A4', Q), ('G4', E), ('F4', D), ('E4', D), ('F4', D), ('G4', H),
    ('D4', D), ('E4', D), ('F4', H), ('E4', D), ('F4', D), ('G4', H),
    ('G4', D), ('A4', Q), ('G4', E), ('F4', D), ('E4', D), ('F4', D), ('G4', H),
    ('D4', H), ('G4', D), ('E4', D), ('C4', H),
]

# ==================== GENERATE ====================
print("=== Generating Music Melodies ===")
print("")

generate_melody(twinkle, "twinkle.wav")
generate_melody(happy, "happy.wav")
generate_melody(wheels, "wheels.wav")
generate_melody(bingo, "bingo.wav")
generate_melody(old_macdonald, "old_macdonald.wav")
generate_melody(itsy_bitsy, "itsy_bitsy.wav")
generate_melody(mary_lamb, "mary_lamb.wav")
generate_melody(row_boat, "row_boat.wav")
generate_melody(abc, "abc.wav")
generate_melody(london_bridge, "london_bridge.wav")

print("")
print("✅ All music melodies generated!")
